### ISSUE TITLE
ES polyfills are not injected

### ISSUE THREAD

**Issue by teppeis** [google/closure-compiler#2881]
repro: https://github.com/teppeis-sandbox/closure-compiler-issue-polyfills
Confirmed in v20180402

### `Symbol` polyfill is not inserted

These expressions don't inject polyfill at all.

```js
alert(Symbol());
alert('iterator' in Symbol);
alert([][Symbol.iterator]());
```

The workaround is `--force_inject_library es6_runtime` or inserting `Array.from([])`, 
but I don't know how to inject Symbol polyfill only.

### `typeof` removes polyfills

These expressions don't inject polyfill at all.

```js
alert(typeof Number.isInteger);
alert(typeof Number.isSafeInteger);
alert(typeof Number.EPSILON);
alert(typeof String.prototype.repeat === "function" && "a".repeat(3));
```

With `--force_inject_library es6_runtime`, polyfills are injected, but SIMPLE optimization removes `Number.EPSILON` and `Number.isSafeInteger` polyfills. (The other polyfills remain)
https://github.com/teppeis-sandbox/closure-compiler-issue-polyfills/blob/master/dist/force-inject.js

If `typeof` is removed, each polyfill is injected.
```js
alert(Number.isInteger);
alert(Number.isSafeInteger);
alert(Number.EPSILON);
alert("a".repeat(3));
```

**Comment by brad4d** [comment]
We currently inject polyfills based on the language features detected by the parser, but it really only notices new syntax, like `let` or `class`.

As part of our efforts to better support features up through `ES_2017` in all parts of the compiler we're planning to start injecting all polyfills that are indicated by the requested input language level.
Before we do that, we want to also make some improvements to the code for removing unused polyfills to make sure we don't end up causing code bloat.

**Comment by teppeis** [comment]
@brad4d Thanks! My understand:

#### Symbol polyfill

`EarlyEs6ToEs3Converter` rewrites Symbol to the polyfill and inserts `initSymbol()`, but it is processed [only if ES6+ syntax features are used](https://github.com/google/closure-compiler/blob/512c02ab38c08b9bdf360935e1af09672481c040/src/com/google/javascript/jscomp/TranspilationPasses.java#L448).

So the current work around for Symbol polyfill is just adding useless `let foo;` and you plan to improve it.

#### `typeof` removes polyfills

It's by design.
https://github.com/google/closure-compiler/commit/076ba4da4327122a2f643f34a5e6095a215a915d introduced `GuardedCallback` to recognize feature detection. It prevents from unintended polyfill insertion. (IMO this behavior is confusing and it's simpler and better to insert polyfill whenever the code includes symbols to polyfill)

The reason why `Number.isSafeInteger` polyfill is removed and `Number.isInteger` polyfill remains is `Number.isSafeInteger` polyfill calls `Number.isInteger` internally. `Number.isSafeInteger` is only used with guard and removed. Only `Number.isInteger` polyfill remains, but no one uses it anymore.

**Comment by teppeis** [comment]
At least `GuardedCallback` is not suitable for constant polyfill like `Number.MAX_SAFE_INTEGER`.
For example, this is not feature detection code.
```js
function isMaxSafeInteger(n) {
  return Number.MAX_SAFE_INTEGER === n;
}
alert(isMaxSafeInteger(1));
```
But the result does not include polyfill because it is recognized as "guarded".
```js
alert(1 === Number.MAX_SAFE_INTEGER);
```

**Comment by brad4d** [comment]
All of these issues will be addressed by #2908 which should get fixed as part of our efforts this quarter to move transpilation of features up through `ES_2017` after all of the checks passes.
